### ISSUE TITLE
If you specify localhost and 127.0.0.1 in different groups, variable not passed from one task to another.

### ISSUE THREAD

**Issue by Denis-Mak** [ansible/ansible#18356]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
Task

##### ANSIBLE VERSION
2.2.0.0

##### CONFIGURATION
Default

##### OS / ENVIRONMENT
CentOS  3.10.0-327.36.1.el7.x86_64

##### SUMMARY
If you specify localhost and 127.0.0.1 in different groups, variable not passed from one task to another.

##### STEPS TO REPRODUCE
File: /src/ansible-bug/environments/vars/hosts
```
[any-other-host]
localhost

[test-vars-host]
127.0.0.1
```

File: /src/ansible-bug/roles/test_vars/tasks/main.yml
```
- command: echo 1
  register: my_var

- debug: var=my_var verbosity=2
```

File: /src/ansible-bug/test_vars.yml
```
- hosts: test-vars-host
  connection: local
  become: yes
  become_user: root
  become_method: sudo
  roles:
    - test_vars
```

Run command:
` ansible-playbook -vvv /src/ansible-bug/test_vars.yml -i /src/ansible-bug/environments/vars/`

##### EXPECTED RESULTS
Echo my_var value. You can see it if you change 'test-vars-host' to 'localhost' or 'any-other-host' to '127.0.0.1'

##### ACTUAL RESULTS
```
ok: [localhost] => {
    "my_var": "VARIABLE IS NOT DEFINED!"
}

```

**Comment by bcoca** [comment]
I cannot reproduce this, i always get the following:

```
ok: [127.0.0.1] => {
    "my_var": {
        "changed": true,
        "cmd": [
            "echo",
            "1"
        ],
        "delta": "0:00:00.001123",
        "end": "2016-11-04 11:33:29.321071",
        "rc": 0,
        "start": "2016-11-04 11:33:29.319948",
        "stderr": "",
        "stdout": "1",
        "stdout_lines": [
            "1"
        ],
        "warnings": []
    }
}

```

**Comment by Denis-Mak** [comment]
[var_test.tar.gz](https://github.com/ansible/ansible/files/572302/var_test.tar.gz)
I have attached src files.
Have you created all files exactly as in example? 
I have just reproduced this bug on another VM created from scratch with 14.04.1-Ubuntu SMP
Just set 'test-vars-host' to IP address instead of 'localhost'

**Comment by ilearnedthisthehardway** [comment]
For whatever it is worth, I have reproduced this behavior with Ansible version 2.1.2.0 on a Fedora system:

```
$ cat /etc/redhat-release
Fedora release 24 (Twenty Four)
$ ansible --version
ansible 2.1.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

**Comment by bcoca** [comment]
My results with your files

```
ansible-playbook test_vars.yml -i environments/vars/hosts -vvv
Using /etc/ansible/ansible.cfg as config file

PLAYBOOK: test_vars.yml *************************************************************************************************************************************************
1 plays in test_vars.yml

PLAY [test-vars-host] ***************************************************************************************************************************************************

TASK [setup] ************************************************************************************************************************************************************
Using module file /home/bcoca/work/ansible-modules-core/system/setup.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: bcoca
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1478536662.3-195993384156720 `" && echo ansible-tmp-1478536662.3-195993384156720="` echo $HOME/.ansible/tmp/ansible-tmp-1478536662.3-195993384156720 `" ) && sleep 0'                                                                                  
<localhost> PUT /tmp/tmpx0VjIF TO /home/bcoca/.ansible/tmp/ansible-tmp-1478536662.3-195993384156720/setup.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/bcoca/.ansible/tmp/ansible-tmp-1478536662.3-195993384156720/ /home/bcoca/.ansible/tmp/ansible-tmp-1478536662.3-195993384156720/setup.py && sleep 0'                                                                                                                                                  
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/bcoca/.ansible/tmp/ansible-tmp-1478536662.3-195993384156720/setup.py; rm -rf "/home/bcoca/.ansible/tmp/ansible-tmp-1478536662.3-195993384156720/" > /dev/null 2>&1 && sleep 0'                                                                                                                 
ok: [localhost]

TASK [test_vars : command] **********************************************************************************************************************************************
task path: /home/bcoca/tmp/src/ansible-bug/roles/test_vars/tasks/main.yml:1
Using module file /home/bcoca/work/ansible-modules-core/commands/command.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: bcoca
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1478536662.52-266070776128628 `" && echo ansible-tmp-1478536662.52-266070776128628="` echo $HOME/.ansible/tmp/ansible-tmp-1478536662.52-266070776128628 `" ) && sleep 0'                                                                               
<localhost> PUT /tmp/tmpaSbXRX TO /home/bcoca/.ansible/tmp/ansible-tmp-1478536662.52-266070776128628/command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/bcoca/.ansible/tmp/ansible-tmp-1478536662.52-266070776128628/ /home/bcoca/.ansible/tmp/ansible-tmp-1478536662.52-266070776128628/command.py && sleep 0'                                                                                                                                              
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/bcoca/.ansible/tmp/ansible-tmp-1478536662.52-266070776128628/command.py; rm -rf "/home/bcoca/.ansible/tmp/ansible-tmp-1478536662.52-266070776128628/" > /dev/null 2>&1 && sleep 0'                                                                                                             
changed: [localhost] => {
    "changed": true,
    "cmd": [
        "echo",
        "1"
    ],
    "delta": "0:00:00.001081",
    "end": "2016-11-07 11:37:42.582825",
    "invocation": {
        "module_args": {
            "_raw_params": "echo 1",
            "_uses_shell": false,
            "chdir": null,
            "creates": null,
            "executable": null,
            "removes": null,
            "warn": true
        },
        "module_name": "command"
    },
    "rc": 0,
    "start": "2016-11-07 11:37:42.581744",
    "stderr": "",
    "stdout": "1",
    "stdout_lines": [
        "1"
    ],
    "warnings": []
}

TASK [test_vars : debug] ************************************************************************************************************************************************
task path: /home/bcoca/tmp/src/ansible-bug/roles/test_vars/tasks/main.yml:4
ok: [localhost] => {
    "my_var": {
        "changed": true,
        "cmd": [
            "echo",
            "1"
        ],
        "delta": "0:00:00.001081",
        "end": "2016-11-07 11:37:42.582825",
        "rc": 0,
        "start": "2016-11-07 11:37:42.581744",
        "stderr": "",
        "stdout": "1",
        "stdout_lines": [
            "1"
        ],
        "warnings": []
    }
}

PLAY RECAP **************************************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0
```

**Comment by ansibot** [comment]
@Denis-Mak ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by samdoran** [comment]
If you still have this problem, please open a new issue.
